arangod/Aql/ExecutionBlockImpl.cpp

```diff
@@ -338,28 +338,55 @@ ExecutionBlockImpl<Executor>::execute(AqlCallStack const& stack) {
   TRI_IF_FAILURE("ExecutionBlock::getOrSkipSome3") {
     THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
   }
-  initOnce();
-  auto res = executeWithoutTrace(stack);
-  if (std::get<0>(res) != ExecutionState::WAITING) {
-    // if we have finished processing our input we reset _outputItemRow and
-    // _lastRange to release the SharedAqlItemBlockPtrs. This is necessary to
-    // avoid concurrent refCount updates in case of async prefetching because
-    // refCount is intentionally not atomic.
-    _outputItemRow.reset();
-    if constexpr (!isMultiDepExecutor<Executor>) {
-      if (!_lastRange.hasValidRow()) {
-        _lastRange.reset();
+  
+  // check if this block failed already.
+  if (_firstFailure.fail()) {
+    // if so, just return the stored error.
+    // we need to do this because if a block fails with
+    // an exception, it is in an invalid state, and all
+    // subsequent calls into it may behave badly.
+    THROW_ARANGO_EXCEPTION(_firstFailure);
+  }
+
+  try {
+    initOnce();
+    auto res = executeWithoutTrace(stack);
+
+    if (std::get<0>(res) != ExecutionState::WAITING) {
+      // if we have finished processing our input we reset _outputItemRow and
+      // _lastRange to release the SharedAqlItemBlockPtrs. This is necessary to
+      // avoid concurrent refCount updates in case of async prefetching because
+      // refCount is intentionally not atomic.
+      _outputItemRow.reset();
+      if constexpr (!isMultiDepExecutor<Executor>) {
+        if (!_lastRange.hasValidRow()) {
+          _lastRange.reset();
+        }
       }
     }
-  }
 #ifdef ARANGODB_ENABLE_MAINTAINER_MODE
-  auto const& [state, skipped, block] = res;
-  if (block != nullptr) {
-    block->validateShadowRowConsistency();
-  }
+    auto const& [state, skipped, block] = res;
+    if (block != nullptr) {
+      block->validateShadowRowConsistency();
+    }
 #endif
-  traceExecuteEnd(res);
-  return res;
+    traceExecuteEnd(res);
+    return res;
+  } catch (basics::Exception const& ex) {
+    if (_firstFailure.ok()) {
```

Comment:
Actually we should never get here if `!_firstFailure.ok()`. How about an assertion instead of the if?

Reply:
done